protobuf/source_reflection.go

```diff
@@ -0,0 +1,71 @@
+package protobuf
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/grpcreflect"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	rpb "google.golang.org/grpc/reflection/grpc_reflection_v1alpha"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/resolver/manual"
+)
+
+var (
+	// ErrorCouldNotDialReflectionServer is an error for when the reflection server could not be reached.
+	ErrorCouldNotDialReflectionServer = errors.New("could not reach reflection server")
+)
+
+// ReflectionArgs are args for constructing a DescriptorProvider that reaches out to a reflection server.
+type ReflectionArgs struct {
+	Caller  string
+	Service string
+	Peers   []string
+	Timeout time.Duration
+}
+
+// NewDescriptorProviderReflection returns a DescriptorProvider DescriptorProvider that reaches out to
+// a reflection server to access filedescriptors.
+func NewDescriptorProviderReflection(args ReflectionArgs) (DescriptorProvider, error) {
+	metadataContext := metadata.NewOutgoingContext(context.Background(),
```

Comment:
Should we apply the configured timeout to our `context.Context` here as well?

Reply:
Maybe, I'm still a bit fuzzy on stream and timeout behaviour. Any timeout that would be set would apply to opening and closing the whole stream for all the reflection related operations. Additionally if we'd use a `context.WithTimeout` here that would set a timeout from the construction of this object rather then from when we actually start making requests.

I'd say that having a timeout on opening a connection to the reflection server is enough, if that passes then the retrieval of the required file descriptors currently doesn't have a timeout. I think that behaviour is similar to not having a timeout on thrift compilation.

Reply:
Yea, that seems reasonable to me. We can always update this later if it ever becomes an issue.